cumulusci/tasks/salesforce/metadata_etl/base.py

```diff
@@ -0,0 +1,262 @@
+
+        self.api_names = list(
+            map(
+                self._namespace_injector,
+                process_list_arg(self.options.get("api_names", [])),
+            )
+        )
+
+    def _get_entities(self):
+        return {self.entity: self.api_names or ["*"]}
+
+    def _transform_entity(self, metadata, api_name):
+        return metadata
+
+    def _transform(self):
+        # call _transform_entity once per retrieved entity
+        # if the entity is an XML file, provide a parsed version
+        # and write the returned metadata into the deploy directory
+
+        parser = PackageXmlGenerator(
+            None, self.api_version
+        )  # We'll use it for its metadata_map
+        entity_configurations = list(
+            filter(
+                lambda entry: any(
+                    [
+                        subentry["type"] == self.entity
+                        for subentry in parser.metadata_map[entry]
+                    ]
+                ),
+                parser.metadata_map,
+            )
+        )
+        if not entity_configurations:
+            raise CumulusCIException(
+                f"Unable to locate configuration for entity {self.entity}"
+            )
+
+        configuration = parser.metadata_map[entity_configurations[0]][0]
+        if configuration["class"] not in [
+            "MetadataFilenameParser",
+            "CustomObjectParser",
+        ]:
+            raise CumulusCIException(
+                f"MetadataSingleEntityTransformTask only supports manipulating complete, file-based XML entities (not {self.entity})"
+            )
+
+        extension = configuration["extension"]
+        directory = entity_configurations[0]
+
+        for api_name in self.api_names:
+            path = self.retrieve_dir / directory / f"{api_name}.{extension}"
+            if not path.exists():
+                raise CumulusCIException(f"Cannot find metadata file {path}")
+
+            transformed_xml = self._transform_entity(
+                elementtree_parse_file(path), api_name
+            )
+            if transformed_xml:
+                parent_dir = self.deploy_dir / directory
+                if not parent_dir.exists():
+                    parent_dir.mkdir()
+                destination_path = parent_dir / f"{api_name}.{extension}"
+                transformed_xml.write(
+                    destination_path,
+                    "utf-8",
+                    xml_declaration=True,
+                    default_namespace=self.namespaces["sf"],
+                )
+
+
+def get_new_tag_index(tree, tag, namespaces):
+    # All top-level tags must be grouped together in XML file
+    tags = tree.findall(f".//sf:{tag}", namespaces)
+    if 0 < len(tags):
+        # Insert new tag after the last existing tag of the same type
+        return list(tree.getroot()).index(tags[-1]) + 1
+    else:
+        # There are no existing tags of this type; insert new tag at the top.
+        return 0
```

Comment:
Bottom might be a less surprising default?

Reply:
The comment is slightly confusing. If there are extent tags of this type (e.g `<relatedList>`), we should append after the last tag of that type. If there are no such tags, we should insert at the head of the file.

This is because MDAPI wants all like-typed tags grouped together.

Reply:
Yes, that's what I understood from the code. If there aren't any existing tags of this type then it doesn't really matter whether the new tag gets added at the top or bottom in order to satisfy MDAPI's grouping requirement, but I'm saying adding them at the bottom would better match what a human would probably do if adding them to the file manually.

Reply:
Ah, pardon my misunderstanding. Yes, I see your point, and I've changed this behavior.